Zephyr native mcumgr taskstat

de-nordic opened the pull request:
The series of commits that make taskstat command Zephyr native with some enhancements.

The PR includes commits:
 - ~~**mgmt/mcumgr/lib: Show taskstat only when THREAD_MONITOR enabled**
   that adds Kconfig dependency that will not show OS_MGMT_TASKTAT options when THREAD_MONITOR is disabled~~
 - **mgmt/mcumgr/lib: Optimize taskstat for Zephyr**
   that moves "taskstat" processing ot os_mgmt.c and removes intermediate, system agnostic, layer when the command  has been served by common code of mcumgr library;
 - **mgmt/mcumgr/lib: Make thread name in taskstat response configurable**
   adds Kconfig option to limit length of thread name that will be included in response
 - **mgmt/mcumgr/lib: Allow signed priorities in taskstat**
   adds Kconfig option allows to send signed priorities, which are native for Zephyr, while previously priorities would have been converted to unsigned numbers;
 - ~~**mgmt/mcumgr/lib: Allow thread ID as thread name in taskstat**
   adds Kconfig option that allows to use thread index, when thread name is disabled, as ID in taskstat response, where tasktat priority has been used previously;~~
 - **mgmt/mcumgr/lib: Thread name selection for taskstat report**
   adds Kconfig manu that allows to choose whether thread name, thread ID/index or thread priority will be used as thread name in taskstat response;
 - **mgmt/mcumgr/lib: Add OS_MGMT_TASKSTAT_STACK_INFO to mcumgr**
   adds Kconfig option that allows whether stack information will be sent in responses; this allows to reduce size of responses and is specially useful when application is built with stack information disabled;
 - **mgmt/mcumgr/lib: Shorten error path in taskstat processing**
   change where lasy evaluation of `||` is used to skip processing of CBOR encoding calls in case if any fails; previously all CBOR encoding calls would be executed anyway.

de-nordic committed:
mgmt/mcumgr/lib: Make thread name in taskstat response configurable

The change adds OS_MGMT_THREAD_NAME_LEN that allows to set length of
thread name that is returned in taskstat response.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Optimize taskstat for Zephyr

The commit changes taskstat code to directly process thread information
from Zephyr structures, instead of translating them to system agnostic
layer, before formatting response.
It also moves the takstat code to os_mgmt.c.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Shorten error path in taskstat processing

The tasktat code filling reposes with use of CBOR has been modified,
utilizing lazy evaluation of C '||' operator, to terminate CBOR
encoding as soon as first error appears.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Allow signed priorities in taskstat

By default the mcumgr taskstat responses send task priorities
as unsigned integers, while Zephyr uses int8_t as priority type.
This commit adds OS_MGMT_TASKSTAT_SIGNED_PRIORITY Kconfig option
that allows to switch to use signed priorities in responses.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Thread name selection for taskstat report

The commit adds Kconfig options and supporting code that allows
to select characteristic that will be used for thread name
in taskstat from:
 - thread name, when THREAD_NAME is enabled;
 - thread ID/index;
 - thread priority.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Add OS_MGMT_TASKSTAT_STACK_INFO to mcumgr

The commit adds OS_MGMT_TASKSTAT_STACK_INFO Kconfig option
that allows to skip, when disabled,  "stksz" and "stkuse" reports,
in "taskstat" command responses.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

lairdjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -251,6 +252,53 @@ config OS_MGMT_TASKSTAT_ONLY_SUPPORTED_STATS
+	  them as unsigned.  Enabling this option will use stigned priorities in
+	  responses, which is more natural for Zephyr.
+	  Enable this option only if your client software is able to properly decode
+	  and accept signed integeres as priorities.
```
*integers

de-nordic replied:
Done

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -45,103 +51,167 @@ os_mgmt_echo(struct mgmt_ctxt *ctxt)
+#else
+	char thread_name[CONFIG_OS_MGMT_TASKSTAT_THREAD_NAME_LEN];
+
+	thread_name[sizeof(thread_name) - 1] = '0';
```
Should this be = 0 literal rather than '0' which is 0x30?

de-nordic replied:
Yes, missing slash. Should have just gone with literal 0.

de-nordic replied:
Done

carlescufi commented:
@lairdjm could you please take another look?

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -45,103 +51,167 @@ os_mgmt_echo(struct mgmt_ctxt *ctxt)
+	int err = 0;
+
+#ifdef CONFIG_THREAD_STACK_INFO
+	stack_size = thread->stack_info.size / 4;
+
+#ifdef CONFIG_INIT_STACKS
+	unsigned int stack_unused;
+
+	if (k_thread_stack_space_get(thread, &stack_unused) == 0) {
+		stack_used = thread->stack_info.size - stack_unused;
```
Using /4 for size but no divisor for used

de-nordic replied:
Good catch.

de-nordic replied:
Done

nvlsianpu commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -268,6 +268,15 @@ config OS_MGMT_TASKSTAT_THREAD_NAME_LEN
+	bool "Enable signed priorities"
+	help
+	  Zephyr uses int8_t as thread priorities, but in taskstat response it encodes
+	  them as unsigned.  Enabling this option will use stigned priorities in
```
stigned -> signed

de-nordic replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -235,6 +235,7 @@ config OS_MGMT_RESET_MS
 	  before performing the reset.  This delay allows time for the mcumgr
 	  response to be delivered.
 
+if THREAD_MONITOR
```
Yes. I will actually rebase this as soon as the mentioned PR gets merged.

de-nordic replied:
Done

nvlsianpu commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -277,6 +277,16 @@ config OS_MGMT_TASKSTAT_SIGNED_PRIORITY
 
+if !THREAD_NAME
+
+config OS_MGMT_TASKSTAT_USE_THREAD_IDX_FOR_NAME
```
I would create a chose of three values for configure what will be transferred under thread name.

de-nordic replied:
This is a little bit complicated because the usage of idx is default while THREAD_NAME is not enabled, but when you enable THREAD_NAME you may still want to use the idx. I was also considering to defaulting the option to y.

de-nordic replied:
Done

nvlsianpu commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -222,7 +207,7 @@ static int os_mgmt_taskstat_read(struct mgmt_ctxt *ctxt)
 	}
 
-	err |= cbor_encoder_close_container(&ctxt->encoder, &tasks_map);
+	err = err | cbor_encoder_close_container(&ctxt->encoder, &tasks_map);
```
Is this unwanted line change?

de-nordic replied:
I have no explanation for the change, so yes unwanted. Will fix.

lairdjm replied:
err from the loop should propagate through to this, should it not?

de-nordic replied:
Yes, but the original line did that anyway.

lairdjm replied:
It does in this view but in the commit view, it shows this as the original:
```
 	err = cbor_encoder_close_container(&ctxt->encoder, &tasks_map);
```
Without the |, not sure why there's a difference in the views

de-nordic replied:
Done

lairdjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -251,6 +252,53 @@ config OS_MGMT_TASKSTAT_ONLY_SUPPORTED_STATS
+	  Enabling this option adds stack information into "taskstat" command
+	  responses, this is default when THREAD_STACK_INFO is selected.
+	  Disable this option only when your client application is able to
+	  process "taskstat" response without the "stksz" and "stkuse" fields.
```
*stksiz

de-nordic replied:
Done

lairdjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -251,6 +252,53 @@ config OS_MGMT_TASKSTAT_ONLY_SUPPORTED_STATS
+	  responses, this is default when THREAD_STACK_INFO is selected.
+	  Disable this option only when your client application is able to
+	  process "taskstat" response without the "stksz" and "stkuse" fields.
+	  It is worth to disable this option when THREAD_STACK_INFO is disabled,
```
*It is worth disabling

de-nordic replied:
Done

de-nordic committed:
mgmt/mcumgr/lib: Remove os_mgmt_config.h

The commit removes os_mgmt_config.h that has been translating Kconfig
options to mgmt internal definitions for constants, and replaces
usage of these constants with direct use of Kconfig options.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>
